crates/uv-configuration/src/dev.rs

```diff
@@ -1,445 +1,350 @@
         };
 
-        let groups = if no_default_groups {
-            // Remove groups specified with `--no-group`.
-            group.retain(|group| !no_group.contains(group));
-
-            Some(GroupsSpecification::Explicit { include: group })
-        } else if all_groups {
-            Some(GroupsSpecification::Include {
-                include: IncludeGroups::All,
-                exclude: no_group,
-            })
-        } else if !group.is_empty() {
-            if matches!(dev, Some(DevMode::Only)) {
-                unreachable!("cannot specify both `--only-dev` and `--group`")
-            };
-
-            // Ensure that `--no-group` and `--group` are mutually exclusive.
-            group.retain(|group| !no_group.contains(group));
-
-            Some(GroupsSpecification::Include {
-                include: IncludeGroups::Some(group),
-                exclude: no_group,
-            })
-        } else if !only_group.is_empty() {
-            if matches!(dev, Some(DevMode::Include)) {
-                unreachable!("cannot specify both `--dev` and `--only-group`")
-            };
-
-            // Ensure that `--no-group` and `--only-group` are mutually exclusive.
-            only_group.retain(|group| !no_group.contains(group));
-
-            Some(GroupsSpecification::Only {
-                include: only_group,
-                exclude: no_group,
-            })
-        } else if !no_group.is_empty() {
-            Some(GroupsSpecification::Include {
-                include: IncludeGroups::Some(Vec::new()),
-                exclude: no_group,
-            })
-        } else {
-            None
-        };
+        Self::from_history(DevGroupsSpecificationHistory {
+            dev_mode,
+            group,
+            only_group,
+            no_group,
+            all_groups,
+            no_default_groups,
+            // This is unknown at CLI-time, use `.with_defaults(...)` to apply this later!
+            defaults: Vec::new(),
+        })
+    }
+
+    /// Helper to make a spec from just a --dev flag
+    pub fn from_dev_mode(dev_mode: DevMode) -> Self {
+        Self::from_history(DevGroupsSpecificationHistory {
+            dev_mode: Some(dev_mode),
+            ..Default::default()
+        })
+    }
 
-        Self { dev, groups }
+    /// Helper to make a spec from just a --group
+    pub fn from_group(group: GroupName) -> Self {
+        Self::from_history(DevGroupsSpecificationHistory {
+            group: vec![group],
+            ..Default::default()
+        })
     }
 
-    /// Return a new [`DevGroupsSpecification`] with development dependencies included by default.
+    /// Apply defaults to a base [`DevGroupsSpecification`].
     ///
     /// This is appropriate in projects, where the `dev` group is synced by default.
-    #[must_use]
-    pub fn with_defaults(self, defaults: Vec<GroupName>) -> DevGroupsManifest {
+    pub fn with_defaults(&self, defaults: Vec<GroupName>) -> DevGroupsManifest {
```

Comment:
It's a bit surprising this returns a new type? I'd expect `with_...` to return `Self`. Perhaps it just needs to be renamed?

Related https://github.com/astral-sh/uv/pull/11224/files#r1942058015

Reply:
Hmm maybe `manifest_defaults` (more renaming for later...)